File: tobas_gcs/src/tobas_gcs/apps/control_system/sections/battery_viewer.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ....gcs import GroundControlStationWidget

import rospy
from overrides import override
from PyQt5.QtWidgets import QLabel

from tobas_std_tools_py.math import remap
from tobas_rqt_tools.widgets import HPositionBarWidget
from tobas_rqt_tools.layouts import FormLayout
from tobas_tools_py.drone import Drone
from tobas_msgs.msg import Battery

from .base_section import BaseControlSystemSectionWidget


class BatteryViewerWidget(BaseControlSystemSectionWidget):
    LABEL = "Battery"

    RANGE_WIDTH = 500
    RANGE_HEIGHT = 30

    def __init__(self, main: GroundControlStationWidget, drone: Drone) -> None:
        super().__init__(main, drone)

        form = FormLayout()
        self._rows.addLayout(form)

        self._voltage_range = HPositionBarWidget()
        self._voltage_range.setFixedSize(self.RANGE_WIDTH, self.RANGE_HEIGHT)
        form.addRow(QLabel("Voltage"), self._voltage_range)

        self._current_range = HPositionBarWidget(fill_range=False)
        self._current_range.setFixedSize(self.RANGE_WIDTH, self.RANGE_HEIGHT)
        form.addRow(QLabel("Current"), self._current_range)

        self._battery_sub = None

    @override
    def update_internal_data_structures(self) -> None:
        self._voltage_range.clear()
        self._voltage_range.set_lower(self._drone.battery.sag_voltage)
        self._voltage_range.set_minimum(self._drone.battery.sag_voltage)
        self._voltage_range.set_maximum(self._drone.battery.max_voltage)
        self._voltage_range.start_timer()

        self._current_range.clear()
        self._current_range.set_minimum(0.0)
        self._current_range.set_maximum(self._drone.battery.max_current)
        self._current_range.start_timer()

        if self._battery_sub is not None:
            self._battery_sub.unregister()
        self._battery_sub = rospy.Subscriber(
            f"{self._drone.drone_name}/battery_filtered", Battery, self._battery_cb, queue_size=1
        )

    def _battery_cb(self, battery: Battery) -> None:
        rate = remap(battery.voltage, self._drone.battery.sag_voltage, self._drone.battery.max_voltage, 0, 100)
        self._voltage_range.set_upper(battery.voltage)
        self._voltage_range.set_text(f"{battery.voltage:.2f} V ({int(rate)} %)")

        self._current_range.set_value(battery.current)
        self._current_range.set_text(f"{battery.current:.2f} A")

File: tobas_gui/tobas_control_system/src/battery_viewer.cpp
#include <format>

#include <tobas_math/core.hpp>
#include <tobas_path_tools/join.hpp>
#include <tobas_qt_tools/widgets/label.hpp>
#include <tobas_qt_tools/layouts/form_layout.hpp>

#include "tobas_control_system/battery_viewer.hpp"

namespace gui
{
namespace gcs
{
BatteryViewerWidget::BatteryViewerWidget(rclcpp::Node::SharedPtr node, const tobas::Drone& drone)
  : node_(node), drone_(drone)
{
  voltage_ = new qt::HPositionBarWidget();
  current_ = new qt::HPositionBarWidget();

  voltage_->setFixedHeight(kBarHeight);
  current_->setFixedHeight(kBarHeight);

  // Layout
  const auto form = new qt::FormLayout();
  form->addVAlignedRow(new qt::Label("Battery Voltage", kLabelPSize), voltage_);
  form->addVAlignedRow(new qt::Label("Battery Current", kLabelPSize), current_);

  setLayout(form);
}

void BatteryViewerWidget::updateInternalDataStructures()
{
  voltage_->clear();
  voltage_->setLower(drone_.battery.sag_voltage);
  voltage_->setMinimum(drone_.battery.sag_voltage);
  voltage_->setMaximum(drone_.battery.max_voltage);

  current_->clear();
  current_->setLower(0.);
  current_->setMinimum(0.);
  current_->setMaximum(drone_.battery.max_current);

  batt_sub_ = ros2::createSubscriber(
    node_, path::join(drone_.name, tobas::kRemoteIfaceTopicNS, tobas::kBatteryTopic), &self::battCb, this);
}

void BatteryViewerWidget::battCb(const tobas_msgs::msg::Battery::ConstSharedPtr& batt)
{
  const auto volt_rate = math::remap(batt->voltage, drone_.battery.sag_voltage, drone_.battery.max_voltage, 0., 100.);
  voltage_->setUpper(batt->voltage);
  voltage_->setText(std::format("{:.2f} V ({:.0f} %)", batt->voltage, volt_rate).c_str());

  current_->setUpper(batt->current);
  current_->setText(std::format("{:.2f} A", batt->current).c_str());

  if (volt_rate < 10.)
    voltage_->setFillColor(Qt::red);
  else if (volt_rate < 20.)
    voltage_->setFillColor(Qt::yellow);
  else
    voltage_->setFillColor(Qt::green);
}
}  // namespace gcs
}  // namespace gui
